### ISSUE TITLE
Add Minimum Description Length MDL Algorithm

### ISSUE THREAD

**Issue by seanlaw** [TDAmeritrade/stumpy#186]
According to the [mSTAMP paper](https://www.cs.ucr.edu/~eamonn/Motif_Discovery_ICDM.pdf) one could use the "minimum description length" (MDL) algorithm to decide how many dimensions to select. It might be helpful to add this to STUMPY.

See Figure 6 and starting in the paragraph right above Figure 6. Additionally, see this [Matrix Profile III paper](https://www.cs.ucr.edu/~eamonn/PID4481999_Matrix%20Profile_III.pdf) which discusses the MDL framework in more detail.

There is also a [background paper by Tanaka et al](https://www.researchgate.net/publication/243134007_Discovery_of_Time-Series_Motif_from_MultiDimensional_Data_Based_on_MDL_Principle) and a [subsequent paper by Hu et al](https://www.researchgate.net/publication/220766245_Discovering_the_Intrinsic_Cardinality_and_Dimensionality_of_Time_Series_Using_MDL)

**Comment by seanlaw** [comment]
@shfa5275 @mexxexx Do you have any ideas how MDL works? I've been looking at some of the papers but it's not clear to me how to implement MDL relative to the mSTAMP paper. Do you have any experience with this?

**Comment by seanlaw** [comment]
@mcyeh We've already implemented mSTOMP but the details on how to apply MDL to the results is vague. Would you mind giving me a high level overview of the MDL steps that we need to take after the multi-dimensional matrix profile (and indices) are computed?

**Comment by shfa5275** [comment]
Hej @seanlaw!

I did not pursue the MDL myself. But the following paper uses the same MDL principle as the 
"Yeh, Chin-Chia Michael, et al. (2017) Matrix Profile VI: Meaningful Multidimensional Motif Discovery. ICDM:565-574"

The paper is entitled:
E. Galbrun, P. Cellier, N. Tatti, A. Termier, and B. Crémilleux. Mining periodic patterns with a MDL criterion. In the proceedings of the European Conference on Machine Learning and Principles and Practice of Knowledge Discovery in Databases (ECML-PKDD'18), September 2018

See the references to the MDL in both the papers, i.e.,:
J. Rissanen. Modeling by shortest data description. Automatica, 14(5):465–471, 1978.
(I was unable to get hold of this paper myself)

The work of E. Galbrun et al., is also implemented in Python. The code is available at:
https://github.com/nurblageij/periodic-patterns-mdl


Other possible useful links are:
https://github.com/PimW/Time-Series-Anomaly-Detection
https://github.com/misilva73/extendedMD

Hope this could be of help!

kind regards,

/S

**Comment by mcyeh** [comment]
@seanlaw 

The MDL step is implemented in [this MATLAB function](https://github.com/mcyeh/mstamp/blob/master/MATLAB/extract_k_motif.m).
The usage of the function is to extract motifs from multi-dimensional matrix profile.

I each iteration, it first find the lowest point in each matrix profiles (1-dimensional mp, 2-dimensional mp, 3-dimensional mp, ..., etc.) which forms the motif candidate set. Than, the MDL principle is used to select the "best" motif out of the candidate set for the current iteration.

Let me know if you have further question.

Thanks,

**Comment by seanlaw** [comment]
I will take a closer look at am certain that I'll have more questions. Thanks, @mcyeh and @shfa5275!

**Comment by seanlaw** [comment]
@mcyeh I've gone through the relevant papers as well as the link to the [Matlab code](https://github.com/mcyeh/mstamp/blob/97e6e180196866da3a0fd9b798032e238f0d76df/MATLAB/unconstrain_search.m) that you had referenced:

```
function [bit_sz, dim_id] = get_bit_save(motif_1, motif_2, n_dim, n_bit)
tot_dim = size(motif_1, 2);
sub_len = size(motif_1, 1);
split_pt = get_desc_split_pt(n_bit);
disc_1 = discretization(motif_1, split_pt);
disc_2 = discretization(motif_2, split_pt);

[~, dim_id] = sort(sum(abs(disc_1 - disc_2), 1), 'ascend');
dim_id = dim_id(1:n_dim);
motif_diff = disc_1(:, dim_id) - disc_2(:, dim_id);
n_val = length(unique(motif_diff));

bit_sz = n_bit * (tot_dim * sub_len * 2 - n_dim * sub_len);
bit_sz = bit_sz + n_dim * sub_len * log2(n_val) + n_val * n_bit;


function disc = discretization(motif, split_pt)
for i = 1:size(motif, 2)
    motif(:, i) = (motif(:, i) - mean(motif(:, i))) / ...
        std(motif(:, i), 1);
end
disc = zeros(size(motif));
for i = 1:length(split_pt)
    disc(motif < split_pt(i) & disc == 0) = i;
end
disc(disc == 0) = length(split_pt) + 1;


function split_pt = get_desc_split_pt(n_bit)
split_pt = norminv((1:(2^n_bit)-1)/(2^n_bit), 0, 1);
```

 but I'm still stuck on some of the MDL details. According to your [Matrix Profile III paper](https://www.cs.ucr.edu/~eamonn/PID4481999_Matrix%20Profile_III.pdf), a time series `T` can be discretized into `b-bit` discrete values in the range [1, 256] via:

```
def discretization(T, b_bit):
    return np.round((T - T.min()) / (T.max() - T.min())) * (2 ** b_bit - 1) + 1
```

However, in your Matlab code the discretization function does not produce the results as described in the Matrix Profile III paper and I can't seem to follow the logic or spot the difference in formulation. Specifically, 

1. I'm not familiar with the inverse normal cumulative distribution function that is being used for `split_pt`. How is this related to MDL?
2. What is `split_pt` for?
3. In your Matlab `discretization` function, I see that you z-normalize (motif) subsequences first and then you do:

```
for i = 1:length(split_pt)
    disc(motif < split_pt(i) & disc == 0) = i;
end
disc(disc == 0) = length(split_pt) + 1;
```

However, it's not clear what is happening here and why it is different from the other discretization formulation. If I were to venture a guess, it feels as if the `get_disc_split_pt` is specific for z-normalized subsequences (i.e., it wouldn't work for subsequences that haven't been z-normallzed like straight non-normalized Euclidean distance) and these split points assume that the subsequence data is normally distributed. Essentially, `split_pt` are bin edges if I'm reading this correctly but it isn't clear how/why these points are chosen. I'm missing the logic here. Then, the subsequence is discretized based on these optimal bin edges or split points. Is that roughly correct? Can you explain what those split points mean? Why are they chosen in this way? Why is this discretization result different from the Python function above that is based on your paper?

What happens if the subsequence were not z-normalized at all? How does this affect the `split_pt`? I ask this because we now support matrix profiles without z-normalization (see `aamp` and related functions in STUMPY). So this will be very important for us.

Thanks in advance!

**Comment by mcyeh** [comment]
@seanlaw 

I will try to answer these questions to the best of my knowledge as I did this 3 years ago.

1. I'm not familiar with the inverse normal cumulative distribution function that is being used for `split_pt`. How is this related to MDL?

In order to apply MDL-principle to time series data, we need to discretize the float point time series to integer time series, and any discretization function works as long as it convert the time series to integer sequence. In other words, both discretization you included in your post works. The matrix profile III paper's discretization is based on [1] while the one used in matrix profile VI might be based on the cardinality reduction techniques used in iSAX [2] (i.e., uneven bin size).

[1] https://www.cs.ucr.edu/~eamonn/ICDM_mdl.pdf
[2] https://www.cs.ucr.edu/~eamonn/iSAX_2.0.pdf

2. What is `split_pt` for?

`split_pt` stores the boundary between each bin in z-score.

3. 1. Is that roughly correct?

Yes

3. 2. Can you explain what those split points mean? 

If I remember correctly, the split points are defined so that each bin has the same area under the curve of normal distribution (i.e., narrower bin around min).

3. 3. Why are they chosen in this way? Why is this discretization result different from the Python function above that is based on your paper?

I don't remember if I further explored in the decision of using an alternative discretization function, but the mentality behind is that this method may be better for z-normalized time series.

3. 4. What happens if the subsequence were not z-normalized at all? How does this affect the `split_pt`?

As `split_pt` stores z-scores, the discretization function won't work if the time series is not z-normalized.

I would suggest using the min max discretization function for time series without normalization with the min and max estimated from the whole time series (instead of the subsequence). It may require further exploration to determine the ideal discretization function under different scenarios.

Hope my answers help

Thanks,

**Comment by seanlaw** [comment]
@mcyeh Thank you so much for your thorough and insightful response! I learned a lot and I now have a much better appreciation and perspective on the problem. I really appreciate you taking the time to help us!

**Comment by seanlaw** [comment]
@mcyeh I've been working on this again and I think I understand most of what is happening but have a few more questions. In the following function:

```
function [bit_sz, dim_id] = get_bit_save(motif_1, motif_2, n_dim, n_bit)
tot_dim = size(motif_1, 2);
sub_len = size(motif_1, 1);
split_pt = get_desc_split_pt(n_bit);
disc_1 = discretization(motif_1, split_pt);
disc_2 = discretization(motif_2, split_pt);

[~, dim_id] = sort(sum(abs(disc_1 - disc_2), 1), 'ascend');
dim_id = dim_id(1:n_dim);
motif_diff = disc_1(:, dim_id) - disc_2(:, dim_id);
n_val = length(unique(motif_diff));

bit_sz = n_bit * (tot_dim * sub_len * 2 - n_dim * sub_len);
bit_sz = bit_sz + n_dim * sub_len * log2(n_val) + n_val * n_bit;
```

1. It looks like `sort(sum(abs(disc_1 - disc_2), 1), 'ascend');` is computing the row-wise distances in "discretized" space and then sorting the distances from smallest to largest in order to pick out the appropriate subspace. Maybe it doesn't matter and the sorted indices are still the same but shouldn't the inner subtraction, `disc_1 - disc_2`, be squared?
2. Why is `n_val` tracking the number of unique values between the `motif_diff`? Instinctively, I would've assumed that we'd track the unique values separately in each dimension (limited to the relevant subspace). So, if the first row in `motif_diff` contained, say, a `[2, 0, 3, 1, 9]` and the second row in `motif_diff` contained a `[0, 1, 4, 1, 5]` then the unique values would be counted separately (i.e., the unique values in each row would be `5` and `4` and then `n_val` would be `4 + 5 = 9`). Maybe I'm not understanding what Matlab is doing (I don't have access to it and am only reading the documentation) but in your formulation, it looks like `n_val` would be `7` and not `9`. 
3. I understand what the variables `tot_dim`, `sub_len`, `n_dim`, and `n_bit` represent and, according to the [MDL framework](https://www.cs.ucr.edu/~eamonn/PID4481999_Matrix%20Profile_III.pdf), we are suppose to compute the space needed for storing the discretized motif and its nearest neighbor (for all dimensions). For a single dimension, one would take a (discretized) nearest neighbor subsequence and subtract it from its (discretized) motif subsequence, count the number of mismatches (i.e., non-zero elements) between them, and then compute the reduced description length (RDL) according to:

```
RDL = (number_of_mismatches) * (log2(sub_len) + n_bit)  # Equation 3
```

This is essentially Equation (3) in the paper above. I follow this much and start to get lost after this as it looks very different from:

```
bit_sz = n_bit * (tot_dim * sub_len * 2 - n_dim * sub_len);
bit_sz = bit_sz + n_dim * sub_len * log2(n_val) + n_val * n_bit;
```

In the first line, `(tot_dim * sub_len * 2 - n_dim * sub_len)`, I understand that `tot_dim * sub_len * 2` accounts for the total space needed for each motif and its nearest neighbor in each dimension (including irrelevant dimensions that aren't in the subspace). I don't fully understand the rationale for why we subtract `n_dim * sub_len` from this. Is it simply subtracting out the space that can be compressed from the space that can't be compressed? And then, in `bit_sz + n_dim * sub_len * log2(n_val) + n_val * n_bit`, you are adding the incompressible bits back together with the compressible subspace dimensions. 

However, following Equation (3) from above, can you explain how you've adapted in order to apply it to one or more dimensions and end up with  `n_dim * sub_len * log2(n_val) + n_val * n_bit`? 

For instance, using our example subspace from above:

```
n_dim = 2
total_dim = 3
sub_len = 5
n_bit = 4
motif_diff = [[2, 0, 3, 1, 9]
              [0, 1, 4, 1, 5]]
```

I would've computed the compressible subspace with something like:

```
sum(motif_diff[0].nonzero()[0]) * (log(sub_len) + n+bit) + sum(motif_diff[0].nonzero()[1]) * (log(sub_len) + n+bit)
```

which simplifies to:

```
(sum(motif_diff[0].nonzero()[0]) + sum(motif_diff[0].nonzero()[1]))* (log(sub_len) + n+bit)
```

I can see that my formulation is quite different from `n_dim * sub_len * log2(n_val) + n_val * n_bit` so I'd appreciate any clarification!

**Comment by seanlaw** [comment]
@mcyeh It's been a while but there has been renewed interest by our users to implement MDL but I still haven't been able to figure it out. Are you able to answer some of the questions [above](https://github.com/TDAmeritrade/stumpy/issues/186#issuecomment-750335386)?

**Comment by Ninimama** [comment]
@seanlaw 
Do you mind if I go through this topic in my free time and see if  I can shed some light onto this (if there is still vagueness)? Since you already go through the materials, could you please let me know what paper/website/slide I should read first?

**Comment by seanlaw** [comment]
@Ninimama Thanks for your inquiry. This has just been resolved and closed by 9fd6c6961ee78528f38613071f48468d9115f1ad.

**Comment by seanlaw** [comment]
@Ninimama Please feel free to start a new Discussion if you would like to talk about it and learn more.
